eth/abc.py

```diff
@@ -72,6 +73,58 @@ class MiningHeaderAPI(rlp.Serializable, ABC):
     timestamp: int
     extra_data: bytes
 
+    @property
+    @abstractmethod
+    def hash(self) -> Hash32:
+        """
+        Return the hash of the block header.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def mining_hash(self) -> Hash32:
+        """
+        Return the mining hash of the block header.
+        """
+
+    @property
+    def hex_hash(self) -> str:
+        """
+        Return the hash as a hex string.
+        """
+        ...
+
+    @property
+    @abstractmethod
+    def is_genesis(self) -> bool:
+        """
+        Return ``True`` if this header represents the genesis block of the chain,
+        otherwise ``False``.
+        """
+        ...
+
+    @abstractmethod
+    def copy(self, *args: Any, **kwargs: Any) -> THeader:
```

Comment:
Why not just `'MiningHeaderAPI'`, so we can drop `THeader`?

Reply:
Because `BlockHeaderAPI` inherits this API and so we don't want to hardcode the return type to `MiningHeaderAPI`.

Reply:
But actually, we don't really *need* it to be on `MiningHeaderAPI` and can move it down to `BlockHeaderAPI` and then hardcode it. This makes the usage slightly nicer because otherwise mypy is asking us for a type annotation at every place where we use `copy` like so:

```python
unused_header: BlockHeaderAPI = header.copy(gas_used=0)
```

If I move it down to `BlockHeaderAPI` we can simply do:

```python
unused_header = header.copy(gas_used=0)
```